# Requirements for IDPCalc Conformer Generator

_under construction_

## Structural Databases

To generate *ab initio* structural conformers of Intrinsically Disordered Proteins,  **IDPCalc Conformer Generator** relies on three _databases_ of three different types:

1. primary sequences
1. secondary structure 
1. phi, psi, omega angles

These databases are generated by software provided within this repository (read further). The source of the information that populates these databases derives from mining the [RCSB PDB DataBank][rcsb] after obtaining a non-redundant set of structures (PDB IDs and chain identifiers) from the [Dubrack Lab: PISCES project][pisces]. We usually selected the *cullpdb_pc90_res2.2_R1.0* subset. However, the process with which the databases are generated allows to seamlessly add other structures that are not contained in the *cullpdb* file or to edit the information in this file freely.

The necessary software resources to download and prepare the databases from a *cullpdb* file are provided by the following Python programs presented in the repository, *at the time of writting not all the listed programs have been written, in those cases links won't work*:

* [idpcalc_pdb_downloader.py][pdbdownloader]
* [idpcalc_pdb2dssp.py][pdb2dssp]
* [idpcalc_pdb_spliter.py][pdbsplitter]
* [idpcalc_pdb_idealizor.py][pdbidealizor]
* [idpcalc_pdb_angle_extract.py][pdbangle]
* [idpcalc_confgen_db_builder.py][dbbuilder]


Upon building, the three databases will contain the specified information (sequence, DSSP and angles) for the different PDBIDs/ChainIDs (*cullpdb* entries), also, the three databases will be aligned equaly so that indexing information can be used transversely across the DBs. A forth database will link the structural information to the PDB/CHAIN identifier. This construction grants traceability throughout the whole process.

The four databases (structure plus indexes) might be text files, binary files or [Python pickled objects][pickle], whichever better fits our needs.

Theses databases will be used by *IDPCalculator Conformer Generator* to build the different conformers.

## Conformer Generation Approach

IDPCalc Conformer Generator will consider different approaches for the conformer building process. All the approach share a **common** feature, that angles have to be given in *chunks* of a predefined (user defined) minimum length.

The different building approaches can be detailed as follows:

### a) Angle Search

#### 1. Build from Loops

Under this approach, conformers are built by using angles from protein chunks of loop regions in folded proteins.

In other words, given a target protein of 100 residues, loop regions are sequentially and randomly selected from the database and their set of angles are used to build the conformer backbone. Different loop regions are concatenated until the length of the target protein is fulfilled.

This is the simplest approach that does not consider residue identity; it only considers sequentially integrity, that is, angles used to build conformers are given as *chunks* of different lengths and NOT as isolated residue angles.

#### 2. Build from sequence identity

The whole primary sequence of a protein ultimately encodes for its global energy landscape and at certain chemical environment; and, the same is valid at a local level, where portions of the primary sequence encode for particular conformational propensities, locally. Therefore, sequence identity matches, between the input IDP sequence and the database, and consequent found angles, might set a favorable initial ground to define an ensemble starting pool.

IDPCalc Conformer Generator will perform string search to identify sequence matches between the input sequence and the database, extracting the angles herein. The new conformer backbone angles are built from the sequence matching angles similarly to what described in [1. Build from Loops](#1-build-from-loops).

We expect this approach to capture regions that encode for multiple secondary structural elements which combination have functional value; for example, small alpha-helices in between extended loops, strong turns, even small beta-beta patterns.

We acknowledge that IDP sequence patterns and residue population differs from that of folded proteins, however, we do expect a significant amount of matching to render this approach meaningful. **The absence of matching** is, within this context, information reach and can be explorer at latter stages of the conformer generation algorithm.

##### Allowing sequence identify mismatch

Because, within some contexts, aminoacids can be replaced without expense of the protein's function (in the case under exploration, conformation landscape), it is important to allow mismatch in sequence identity searches. A parameter can be defined to allow the user the specify the maximum amount of mismatch allowed for angle *chunks*; recall the minimum size of residues (angles) *chunks* is also defined by the user.

##### Mapped mismatch

In addition to binary mismatch, i.e. absolute matches or mismatches, a mismatch map needs to be implemented in the knowledge base of IDPCalc Conformer Generator. Mismatch maps specify interchangeable residues pairs: in other words, allowed residues mismatches. These should not be considered as pure matches, neither as pure mismatches.

A parameter will allow the user to specify the level of matching/mismatching allowed.

Usage of user defined matching maps will be implemented, where user provided maps update the builtin maps.

#### 3. Incorporation of sequence motifs

It is well known and acknowledged that sequence motifs directly encode for biological functionality, however, identification of such motifs in native protein sequences is not trivial. It is **not** the current aim of Conformer Generator to search for unknown motifs in intrinsically disordered protein sequences. But, following the lines of the previous steps, there is a need to implement motif search within the wide scope of conformer generation, in order to pull from the database those motif-related angle chunks and use them in the building steps.  

Technically, motif search can be incorporated in the [mismatch search](#allowing-sequence-identity-mismatch) algorithm because it implies the same technical problems: *mismatched identity*. However, there are two considerations that make motif search diverge from *identity mismatch*, these are:

1. some known motifs require high percentage of mismatch, percentages which wouldn't, *a priori*, be allowed in sequence identity searches, and;
2. the identity matches within motifs are not matches of *random* residues, these are actually well defined matches of commonly repeated residues or residue patterns.

These considerations force the implementation of user-driven motif search in IDPCalc Conformer Generator. In other words, allowing the user the possibility to specify motifs to be searched and queried from the database.

##### predefined motifs

Additionally to user input, known IDP motifs can be predefined in the program's knowledge base. Such motifs would be search in both the input sequence and the database and used if a match is found.

### b) Building the Conformers

.

## User Interface

User interface should be that of a shell command-line interface. Graphical interfaces can be considered at a latter stage.

### Usage

The following describes examples of usage:

```
./conformer_generator \
    -n NUMBER_OF_CONFORMERS_TO_GENERATE \
    -m MINIMUM_PROTEIN_CHUNK_SIZE \
    -x MAXIMUM_ALLOWED_MISMATCH \
    -l MAXIMUM_%_OF_ANGLES_FROM_LOOP_REGIONS \ ?? still considering
    -h MAXIMUM_%_OF_ANGLES_FROM_HELICES \ ??
    -e MAXIMUM_%_OF_ANGLES_FROM_SHEETS \ ??
    -o FOLDER_WHERE_TO_DEPOSIT_THE_STRUCTURES, defaults to CWD
    

```

[pdbdownloader]: https://github.com/joaomcteixeira/IDPCalcPDBDownloader/blob/master/idpcalc_pdb_downloader.py
[pdb2dssp]: https://github.com/joaomcteixeira/IDPCalcPDBDownloader/blob/master/idpcalc_pdb2dssp.py
[pdbsplitter]: https://github.com/joaomcteixeira/IDPCalcPDBDownloader/
[pdbidealizor]: https://github.com/joaomcteixeira/IDPCalcPDBDownloader/
[pdbangle]: https://github.com/joaomcteixeira/IDPCalcPDBDownloader/
[dbbuilder]: https://github.com/joaomcteixeira/IDPCalcPDBDownloader/
[rcsb]: https://www.rcsb.org/
[pisces]: http://dunbrack.fccc.edu/Guoli/pisces_download.php
[pickle]: https://docs.python.org/3/library/pickle.html
